### PR TITLE
Allow D3D12Device to use an existing device handle

### PR DESCRIPTION
**Pull request by lucy96chen**
Changes:
- Added a new field `existingDeviceHandle` to `IDevice::Desc`, which is 0 by default to indicate that a device does not yet exist and will need to be created
- Modified `D3D12Device::initialize` to check if `desc.existingDeviceHandle` is 0. If not 0, then it stores the value in `m_device` instead of creating a new device through `_createDevice`

@csyonghe Here are the changes I made based on the conversation and notes from yesterday.

### CODE REVIEW

File: tools/gfx/d3d12/render-d3d12.cpp [hunk 1]
```diff
@@ -4287,37 +4287,45 @@ Result D3D12Device::initialize(const Desc& desc)
...
-    if (!m_deviceInfo.m_adapter)
+        // Set the device
+        m_device = m_deviceInfo.m_device;
+    }
+    else
     {
-        // Couldn't find an adapter
-        return SLANG_FAIL;
+        // Set the existing device stored in desc
+        m_device = (ID3D12Device*)desc.existingDeviceHandle;
```
Comment (csyonghe) [inline tools/gfx/d3d12/render-d3d12.cpp:4326]: I think you'll also need to assign to pointer to `m_deviceInfo.m_device`.

File: slang-gfx.h [hunk 2]
```diff
@@ -1424,6 +1424,8 @@ class IDevice: public ISlangUnknown
         DeviceType deviceType = DeviceType::Default;
         // Name to identify the adapter to use
         const char* adapter = nullptr;
+        // The device handle. If 0, a device will need to be created.
```
Comment (csyonghe) [inline slang-gfx.h:1427]: This is OK for now, but for vulkan we may need to pass in both a `VkInstance` and `VkDevice`, so perhaps we can use something like:
```
uint64_t existingDeviceHandles[2];
```
Or even better we can define a `struct` type:
```
struct ExistingDeviceHandle
{
      uint64_t values[2];
};
```

**Review by csyonghe — APPROVED** [review]
Looks good to me. We can either implement the suggested changes in this PR or in your next change.

File: tools/gfx/d3d12/render-d3d12.cpp (hunk 1 above)
Comment (lucy96chen) [reply tools/gfx/d3d12/render-d3d12.cpp:4326]: Yeah, I missed that. Done.

File: slang-gfx.h (hunk 2 above)
Comment (lucy96chen) [reply slang-gfx.h]: Done. I'll leave the helpers you mentioned for later.

File: slang-gfx.h [hunk 3]
```diff
@@ -1424,6 +1424,13 @@ class IDevice: public ISlangUnknown
         DeviceType deviceType = DeviceType::Default;
         // Name to identify the adapter to use
         const char* adapter = nullptr;
+        // Device handles (if they already exist)
+        struct existingDeviceHandles
```
Comment (csyonghe) [inline slang-gfx.h:1428]: We do have a convention that type names are BigCamel. So here should be ExistingDeviceHandle.

**Review by csyonghe — APPROVED** [review]
Would be good to follow the BigCamel naming convention for type names.

File: slang-gfx.h (hunk 3 above)
Comment (lucy96chen) [reply slang-gfx.h]: Derp, fixed.

**Review by csyonghe — APPROVED** [review]

**Pull request merged**